feat(maidr.show): support py-shiny renderer

krishnanand5 opened the pull request:
This PR adds support for maidr on py-shiny through a custom renderer called `render_maidr`. This decorator internally calls the newly created `maidr.render` function to transform a plot of the matplotlib type to a maidr object and envelops it as an iframe srcdoc for display on a py-shiny web dashboard.

Closes [#65]

jooyoungseo commented on maidr/custom_renderers/pyshiny_renderer.py:
Add type annotations for all the functions

jooyoungseo replied:
Change the module name to `shiny_renderer`

jooyoungseo commented on maidr/custom_renderers/pyshiny_renderer.py:
```diff
@@ -0,0 +1,48 @@
+            html_content = rendered_dict.get("html", "")
+            dependencies = rendered_dict.get("dependancies", [])
+
+            # Create a string to hold all dependency tags
+            dependency_tags = ""
+            for dep in dependencies:
+                if dep.get("type") == "script":
+                    dependency_tags += f'<script src="{dep["src"]}"></script>'
+                elif dep.get("type") == "stylesheet":
+                    dependency_tags += f'<link rel="stylesheet" href="{dep["href"]}">'
+
+            # Combine dependencies and plot HTML
+            full_html_content = f"""
+            <!DOCTYPE html>
+            <html>
+            <head>
+                {dependency_tags}
+            </head>
+            <body>
+                {html_content}
+            </body>
+            </html>
+            """
+
```
any reasons why you don't use htmltools package and its tag methods instead of the direct string literals for html tags?

jooyoungseo replied:
> During the initial development of the feature, I was able to load the entire maidr object on to the shiny dashboard component but was unable to make use of maidr's interactiveness. Upon analyzing the scripts, I realized that maidr's javascript objects were being loaded on to the webpage but shiny was preventing me from using it.
> 
> In order to have control over the HTML component I am feeding into iframe, I decided to use generic string structures for the HTML tags instead of html tools.
> 
> With the current implementation as a viable fallback, I will begin the migration to htmltools for this section.

Have you taken a look at this method by chance?: https://shiny.posit.co/py/api/core/Htmltools.html

jooyoungseo commented on example/py-shiny/example_pyshiny_scatter.py:
```diff
@@ -0,0 +1,38 @@
+# Load the datasets
+penguins = sns.load_dataset("penguins")
+iris = sns.load_dataset("iris")
+
+# Define the UI for the application
+app_ui = ui.page_fluid(ui.row(ui.column(12, ui.output_ui("scatterplot"))))
+
+
+# Define the server logic
+def server(input, output, session):
+    def create_scatterplot():
+        fig, ax = plt.subplots(figsize=(10, 6))
+        s_plot = sns.scatterplot(
+            data=iris, x="sepal_length", y="sepal_width", hue="species", ax=ax
+        )
+        ax.set_title("Iris Sepal Length vs Sepal Width")
+        ax.set_xlabel("Sepal Length")
+        ax.set_ylabel("Sepal Width")
+        return s_plot
+
+    @output
+    @render.ui
+    def scatterplot():
+        return render_maidr(create_scatterplot)()
+
+
```
This is too simple. Demonstrate the example Shiny app where users can choose two different numeric variables (one for x; the other for y) from the iris dataset and update the plot reactively against the variable change.

krishnanand5 replied:
Sure, Professor. I will create a more complex example to demonstrate the feature more intuitively.

jooyoungseo commented on maidr/custom_renderers/pyshiny_renderer.py:
```diff
@@ -0,0 +1,48 @@
+from matplotlib.axes import Axes
+from matplotlib.figure import Figure
+import maidr
+from shiny import ui
+from typing import Any
```
Add the new shiny dependency via `peotry add shiny` to update the `project.toml`

jooyoungseo commented:
@krishnanand5 I think you have accidentally closed this PR. Please reopen.

krishnanand5 commented:
Greetings Professor,

In an attempt to resolve a merge conflict with poetry.lock, I synced my fork and it caused this unforeseen change. I am rectifying my local repo and will be making a commit shortly with the updated scripts.

Regards,
Krishna Anandan Ganesan.

jooyoungseo commented:
Aw, I see. Thanks for letting me know!

krishnanand5 commented:
Greetings Professor,

I have addressed all the initial concerns in the respective sections and would like to discuss one change that I have implemented in the `shiny_renderer`. 

As shiny provides complete freedom to the user to define the cosmetics of a component that they are creating, the same applies to plot objects as well. In order to view a maidr plot, we require the component area to be slightly bigger so that the user can experience the interactiveness without having to use scrollbars - in other words, overflow of the web component is a factor to consider/document. 

For now, I have modified the render to expect a width and height parameter during function invocation. If you feel we should handle this in a better way, please do suggest any measures and I will begin working on them. I am investigating `htmltools` to identify any support that could help us achieve this.

Regards,
Krishna Anandan Ganesan.

jooyoungseo commented:
@krishnanand5 I cannot execute the new example:

![image](https://github.com/user-attachments/assets/e0711b65-d17e-4511-86dc-9b97bf32fc18)

jooyoungseo commented on example/py-shiny/example_pyshiny_reactive_scatter.py:
```diff
@@ -0,0 +1,57 @@
+from maidr.custom_renderers.shiny_renderer import render_maidr
+import matplotlib.pyplot as plt
+import pandas as pd
```
Why are you importing `pandas`? Is this used?

krishnanand5 replied:
`pandas` was used with the older example but I did not have that removed for the new scatterplot. It has been removed in the latest commit.

jooyoungseo commented on maidr/custom_renderers/shiny_renderer.py:
```diff
@@ -0,0 +1,63 @@
+from matplotlib.axes import Axes
+from matplotlib.figure import Figure
+import maidr
+from shiny import ui
+from htmltools import tags
+from typing import Any, Optional, Callable, TypeVar, Union, List
+
```
Are `matplotlib` and `htmltools` used after you import?

krishnanand5 replied:
Yes both imports are not used after they are imported. Remnants from the older version of the `shiny_renderer` and they have been removed in the latest commit.

jooyoungseo commented on maidr/custom_renderers/shiny_renderer.py:
Some docstrings exceed our line length limit (our default 88 chars). If you use VSCode, you can easily see the linter errors.

krishnanand5 replied:
I was using `Pylint` on my vscode which was set to a default of `130 characters`.  I realized now that `flake8` is the repository's lint module and hence changed the linter on my local vscode as well. Such issues should not arise in future.

jooyoungseo commented:
@SaaiVenkat and @dakshpokar -- Please review this PR as well.

jooyoungseo reviewed (approved):
LGTM!

dakshpokar commented on maidr/custom_renderers/shiny_renderer.py:
```diff
@@ -0,0 +1,67 @@
+
+    def wrapper(
+        *args: Any,
+        width: Optional[str] = None,
```
Hi @krishnanand5,
Can you automatically resize the iframe if height and width are not passed as arguments? Please feel free to refer to #64 for guidance. Considering this library's target audience, we should try to ensure the interface is as user-friendly and accessible as possible.

Apart from that the PR looks clean, good work!

As for the [comment](https://github.com/xability/py_maidr/pull/64/files#r1720523929) that you left on #64 - My observation is that Python notebooks typically have a considerable level of flexibility when it comes to handling cell rendering, which makes suitable style changes possible. Overflow characteristics, however, can be activated if pre-defined styles are used, particularly if the content is larger than the cell's dimensions.

To handle this, you may want to try utilizing Inline-JS/CSS to more accurately regulate the overflow behavior.

SaaiVenkat replied:
- Since iframe seems to addressed here as well, I would suggest to merge this PR after #64 . 
- Also, please create a package called 'widget' under 'maidr' package (maidr/widget/shiny.py), and move the contents there. We'll be extending this further to support GUI widgets in the future.
- We are using the htmltools ui underneath, and shiny uses the same. This annotation duplicates many of the functionalities from shiny as well as from htmltools. This could be simplified by the following

```
from __future__ import annotations

from shiny.render import ui
from shiny.types import Jsonifiable

import maidr


class render_maidr(ui):
    async def render(self) -> Jsonifiable:
        value = await self.fn()
        if value is None:
            return None

        value = maidr.render(value)
        rendered = await self.transform(value)
        return rendered
```

krishnanand5 replied:
The requested changes have been addressed in [in this commit](https://github.com/xability/py_maidr/commit/4fdfcd119a90c7470430d0e2f158f431cfec717d).

SaaiVenkat commented on example/py-shiny/example_pyshiny_reactive_scatter.py:
```diff
@@ -0,0 +1,57 @@
+
+    @output
+    @render.ui
+    def reactivebarplot():
```
Could `@render_maidr` be used to render the example? If so, please update this to reflect the same.

SaaiVenkat commented on example/py-shiny/example_pyshiny_reactive_scatter.py:
```diff
@@ -0,0 +1,57 @@
+
+# Define the server
+def server(input, output, session):
+    @reactive.Calc
```
`reactive.Calc` is mainly used to perform mathematical intensive calculations. May I know why this has been used here?

SaaiVenkat commented on example/py-shiny/example_pyshiny_reactive_scatter.py:
```diff
@@ -0,0 +1,57 @@
+        ax.set_ylabel(input.y_var().replace("_", " ").title())
+        return s_plot
+
+    @output
```
This plot has been injected into the shiny through `ui.output_ui`. Why `@output` is required?

SaaiVenkat commented on maidr/maidr.py:
```diff
@@ -10,6 +11,14 @@
 
+def render(
+    plot: Any, *, lib_prefix: str | None = "lib", include_version: bool = True
+) -> RenderedHTML:
```
In order to incorporate the simplified rendering we can return the `Tag` object instead of the `RenderedHTML`.

```
def render(plot: Any) -> Tag:
    ax = FigureManager.get_axes(plot)
    maidr = FigureManager.get_maidr(ax.get_figure())
    return maidr.render()
```

And update core/maidr.py to render the `Tag`
```
    def render(self) -> Tag:
        return self._create_html_tag()
```

krishnanand5 replied:
Thank you @SaaiVenkat, for these suggestions. These changes have been incorporated and have made the module more compact

SaaiVenkat commented on example/py-shiny/example_pyshiny_reactive_scatter.py:
```diff
@@ -0,0 +1,57 @@
+
+# Define the server
+def server(input, output, session):
+    @reactive.Calc
```
`reactive.Calc` is mainly used to perform mathematical intensive calculations. May I know why this has been used here?

krishnanand5 replied:
Yes, `reactive.calc` is residual from development. Will remove this from the example

SaaiVenkat commented on example/py-shiny/example_pyshiny_reactive_scatter.py:
```diff
@@ -0,0 +1,57 @@
+        return s_plot
+
+    @output
+    @render.ui
```
This plot has been injected into the shiny through `ui.output_ui`. Why `@output` is required?

krishnanand5 replied:
`@output` is for rendering reactive output on the UI but since `@render.ui` is sufficient in our case to render the generated HTML on the shiny app, `@output` is redundant. Will have this removed.

SaaiVenkat commented on example/py-shiny/example_pyshiny_reactive_scatter.py:
```diff
@@ -0,0 +1,57 @@
+    @output
+    @render.ui
+    def reactivebarplot():
+        return render_maidr(create_reactivebarplot)(width="100%", height="700px")
```
Could `@render_maidr` be used to render the example? If so, please update this to reflect the same.

krishnanand5 replied:
As `@render_maidr` is a decorator, we can invoke this directly.

krishnanand5 commented:
Greetings @SaaiVenkat,

Following are the list of changes I have done in the recent commits:

- Moved `shiny_renderer.py` to `maidr/widget/shiny.py`.
- Updated the example plot to use only `render_maidr` and not use default shiny renderers like `@output` and `@render.ui`.
- `@reactive.calc` served no purpose for the provided example script and has been removed.
- `shiny.py` has been simplified to use the `render` of maidr.
- As discussed, I have removed the conditional check in `_inject_plot` to allow shiny to utilize the same iframe generator as the one @dakshpokar developed for Interactive Python notebooks for Shiny as well.

I did check possible collateral effects of removing the conditional check but turns out only the interactive python module and shiny module are currently leveraging it for iframes. Please let me know if the changes made are as per expectations and if there are any further changes you would like on this module.

Regards,
Krishna Anandan Ganesan.

jooyoungseo commented on maidr/widget/shiny.py:
```diff
@@ -0,0 +1,29 @@
+import maidr
+
+
+class render_maidr(ui):
```
We use `PascalCase` for class name in our package. why do you use `snake_case` here? Is this the Shiny convention?

krishnanand5 replied:
I checked Shiny and there are no such conventions. I have renamed the class name to adhere to `PascalCase`.

SaaiVenkat commented on maidr/widget/shiny.py:
```diff
@@ -0,0 +1,29 @@
+import maidr
+
+
+class RenderMaidr(ui):
```
@krishnanand5 shiny breaks the conventional Pascal case for classes, since this will be used as a decorator, like `@render_maidr`. There have followe this convention for all their renderers. So, we should also follow their convention.

cc: @jooyoungseo

krishnanand5 replied:
Thank you for the confirmation @SaaiVenkat. I thought our repository's naming convention superseded supporting libraries but you are right. Better to follow the naming convention of the supporting libraries. I have reverted the class name to `snake_case`
